---
title: "moire"
output: html_document
author: "Kathryn Murie, Maxwell Murphy"
date: "7-Jan-25"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(kableExtra)
```

<br/>
<br/>

## Summary sheet
```{r, echo=FALSE, warning=FALSE}
# read in summary sheet from file
df_summary <- read.csv("summary_sheet.csv", header = FALSE)

# produce formatted table
df_summary |>
  kbl(format = "html", col.names = NULL, table.attr = "style='width:50%;'", escape = FALSE) |>
  kable_styling(position = "left", bootstrap_options = c("striped", "hover", "condensed"))
```

## Purpose

The *moire* (Multiplicity Of Infection and allele frequency REcovery) tool can be used to estimate allele frequencies, MOI, and within-host relatedness from genetic data subject to experimental error. It utilises a Markov Chain Monte Carlo (MCMC) based approach to Bayesian estimation and can take both polyallelic and SNP data as inputs. This tool also introduces a new metric called effective MOI (eMOI), which combines MOI and within-host relatedness into a unified and comparable measure of genetic diversity.  

## Existing resources

- The [*moire* website](https://eppicenter.github.io/moire/index.htm) provides basic usage instructions
- Within the [*moire* website](https://eppicenter.github.io/moire/articles/mcmc_demo.html) there is a more in depth tutorial using simulated genotyping data. 

## Citation
```{r citation}
citation('moire')
```