---
title: "Running Moire Example"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(kableExtra)
options(warn = -1)
```

## The data

In this analysis we will be using the SNP barcode data from the sanger 100 SNP Plasmodium falciparum barcode [(Chang et al. 2019)]( https://doi.org/10.7554/eLife.43481). 
This was generated by subsetting the WGS data (also described within the [Data section](../../website_docs/data_description.qmd) of this website) and we will be using the data from the DRC. 
See the description in the [Data section](../../website_docs/data_description.qmd) for more details on this dataset.

In this tutorial we will use `PGEhammer` to convert data from the VCF format to the format required by MOIRE. To install the package run the following command

```{r install_packages, warning=FALSE, message=FALSE, error=FALSE}
# Install PGEhammer in R:
install.packages('PGEhammer', repos = c('https://plasmogenepi.r-universe.dev', 'https://cloud.r-project.org'))
```

Now we can import libraries that we will need. 

```{r, warning=FALSE, message=FALSE, error=FALSE}
library(tidyverse)
library(here)
library(vcfR)
library(PGEhammer)
```

`MOIRE` requires input data in either long or wide format, which can be loaded using the `load_long_form_data()` or `load_delimited_data()` functions, respectively. The long format represents data with each observed allele for each sample on a separate row, while the wide format uses a row for each sample, and a separate column for each genomic target, with each cell containing a string indicating the collection of alleles that are observed. In the following steps, we will convert the Variant Call Format (VCF) data to the required long format using the function vcf2long from PGEhammer.

```{r load_data}
# Load the vcf  
vcf <- read.vcfR('../../data/snp_barcode/sangerBarcode_SNP_INDEL_Pf3D7_ALL_v3.combined.filtered.vqslod6.biallelic_snp.DRCongo.vcf.gz')
# Convert to long format 
df <- vcf2long(vcf)

head(df) |>
  kable() |>
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```

Prior to executing `MOIRE`, it is advisable to remove alleles that are likely to be erroneous. In the following step, we apply a filter to remove alleles with a read count below 10, however, how you filter your data should be informed by prior knowledge and experience with your data.
```{r filter}
df <- df |>
  dplyr::filter(read_count >= 10)
```

For the purpose of this tutorial we will subset the data to just include chromosome 1
```{r subset}
subset_df <- df[grepl('^Pf3D7_01', df$locus), ]
```

## Running the MCMC 

Now that we have the data in the correct format we will run the Markov chain Monte Carlo (MCMC) with the default parameters.

```{r mcmc, cache = TRUE}
data <- moire::load_long_form_data(subset_df)
mcmc_results <- moire::run_mcmc(data, is_missing = data$is_missing, verbose = FALSE)
```

## Summarising the results

From the MCMC results we can produce estimations for each sample and summarise the results. 

First we will look at COI using `summarize_coi`.

```{r}
# Estimate the COI for each sample
coi_summary <- moire::summarize_coi(mcmc_results)

head(coi_summary) |>
  kable() |>
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```

This dataframe includes summaries of both the posterior distribution of COI for each biological sample and the naive estimates.

Below we use the `summarize_he` function to summarise locus heterozygosity from the posterior distribution of sampled allele frequencies.
```{r}
he_summary <- moire::summarize_he(mcmc_results)

head(he_summary) |>
  kable() |>
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```

Using `summarize_allele_freqs` we can look at individual allele frequencies
```{r}
allele_freq_summary <- moire::summarize_allele_freqs(mcmc_results)

head(allele_freq_summary) |>
  kable() |>
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```

The `summarize_relatedness` function provides a dataframe of within-host relatedness
```{r}
relatedness_summary <- moire::summarize_relatedness(mcmc_results)

head(relatedness_summary) |>
  kable() |>
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```

The *moire* tool introduces a new metric called effective MOI, which adjusts for within-host relatedness. A detailed description of this metric and how to interpret it can be found [here](https://eppicenter.github.io/moire/articles/mcmc_demo.html#what-is-effective-coi).
```{r}
effective_coi_summary <- moire::summarize_effective_coi(mcmc_results)

head(effective_coi_summary) |>
  kable() |>
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))

```

## Summary

In summary, *moire* estimates allele frequencies, MOI, and within-host relatedness. We have shown how to generate basic results from a VCF. `MOIRE` has extensive [documentation](https://eppicenter.github.io/moire/index.html), including more details on [other functionality](https://eppicenter.github.io/moire/articles/mcmc_demo.html#estimating-allele-frequencies) available within the tool and a [tutorial](https://eppicenter.github.io/moire/articles/mcmc_demo.html#estimating-allele-frequencies) validating the outputs using simulated data.