---
title: "Give this whatever name you want"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(kableExtra)
```

## The data

To perform the analysis estMOI requires:

* A bam file which should be sorted
* A vcf file which can either be zipped or unzipped
* The reference genome

For the purpose of hosting the data we have provided bam and vcf files containing data for the 
_csp_, _celtos_ and _ama1_ genes.

We now need to download the reference genome. You can download this from plasmodb by clicking on 
[this link](https://plasmodb.org/common/downloads/release-66/Pfalciparum3D7/fasta/data/PlasmoDB-66_Pfalciparum3D7_Genome.fasta)

## Running the tool

Let's first run the tool on a sample with a known MOI > 1. Find the data in XXX

Before we can run estMOI we need to index the reference genome. This can by done with

``` bash
samtools faidx PlasmoDB-66_Pfalciparum3D7_Genome.fasta
```

Now let's run the pipeline with default parameters. The three positional arguments should be:

1. The bam file 
2. The VCF file
3. The reference genome

You can also customise the output file prefix with `--out`

``` bash
estMOI bams/PG0402-C.sorted.bam bams/PG0402-C.vcf PlasmoDB-66_Pfalciparum3D7_Genome.fasta --out PG0402-C
```

After running it will print a summary of the analysis to the terminal and you should see the output below:

``` bash
#	RUNNING estMOI version 1.03
#	................
#	PRINT ESTIMATES...
#	MOI	Count	%Total
#	1	319	66.88
#	2	158	100.00	MOI-estimate
	#DONE MOI-estimate using bams/PG0402-C.sorted.bam
```

### Output description

Running estMOI will produce two output files with the prefix that you specified before.
Additionally, the values of the `--maxsnp`, `--mindis`, `--maxdis` and `--minhap` parameters will also 
be present in the output file names.

For example, you should have the files PG0402-C.moi.3.10.500.3.log and PG0402-C.moi.3.10.500.3.txt 
in your directory. 


#### Log file

The file ending in .log contains the individual combinations of all haplotypes that were found.
For each haplotype it will give you the locations of the SNPs together with the individual haplotype 
combinations. For example, here is an excerpt form the log file that contains this information for one 
haplotype:

``` bash
bams/PG0402-C.sorted.bam	Pf3D7_03_v3	221435 221457 221554	2
	# bams/PG0402-C.sorted.bam Pf3D7_03_v3 221435 221457 221554 Hapotype:	 T A C	48
	# bams/PG0402-C.sorted.bam Pf3D7_03_v3 221435 221457 221554 Hapotype:	 T G C	10
```

#### Txt file

The result file will contain the the same information as printed to the terminal.
It lists the numbers of haplotypes and for each MOI and provides an estimate of the MOI.
The the example below there are 319 locus with a single haplotypes and 158 with two haplotypes.

``` bash
#MOI	Count	%Total
1	319	66.88
2	158	100.00	MOI-estimate
```



## Summary

It's a good idea to round off with a summary of what we did.

PS, note that the name of the tutorial on the PGEforge website can be whatever you want - it does not have to match the name of this tutorial.