---
title: "isoRelate tutorial for generating Identity-by-decent networks: vcf to network"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(kableExtra)
library(isoRelate)
library(SeqArray
```

## Useful tutorial examples

If you are starting from ped and map files, there is a very clear and comprehensive tutorial `introduction.Rmd` created by isoRelate's author, Lyndal Henden, that uses isoRelates in built data to work through the package functions.
https://github.com/bahlolab/isoRelate/blob/master/vignettes/introduction.Rmd

The data used in the vignette tutorial is a _Plasmodium falciparum_ dataset from Papua New Guinea (PNG), generated by the MalariaGEN Consortium (https://www.malariagen.net/projects/pf3k). When you load isoRelate you are able to trial the functions and visualise how your data should be formated.

## New tutorial

This tutorial focuses on data from the processed vcf stage and works through the process to generate appropriate map and ped files from vcf data within R.


## The data

The data in this tutorial is the SNP barcoding SpotMalariapfPanel_simData_sanger100.vcf.gz (https://mrc-ide.github.io/PGEforge/data/). Please refer to the [main Data tab](https://mrc-ide.github.io/PGEforge/website_docs/data_description.html) for further details on this dataset.

There will be errors if the vcf files include indels. During you're vcf filtering steps, remove indels.

## Step 1

Process your vcf files into pedmap format.

The input for isoRelate is unpahsed genotype data for SNPs in PLINK PED and MAP format (http://www.cog-genomics.org/plink2). You can choose to process your vcf files using PLINK prior to using this workflow. A R script has been developed to improve this process.



```{r vcf_file_to_pedmap_format}
showfile.gds(closeall = TRUE)
vcf_to_gds("pf7.PNG.radish.vcf.gz", "pf7.PNG.radish.gds")  

gds_to_pedmap("pf7.PNG.radish.gds", "my_map.map", "my_ped.ped")

```

Once you have generated the ped and map files, the next step is to generate ibd metrics.
## Step 2

```{r ibd_segments}
# load your ped and map files 
pedmap <- list()
pedmap[[1]] <- read.delim("ped_file.ped", 
                          stringsAsFactors = FALSE, header=FALSE)
pedmap[[2]] <- read.delim("map_file.map", stringsAsFactors = FALSE,
                          header=FALSE)

my_genotypes <- isoRelate::getGenotypes(ped.map = pedmap,
                                        reference.ped.map = NULL,
                                        maf = 0.01,
                                        isolate.max.missing = 0.3,
                                        snp.max.missing = 0.3,
                                        chromosomes = NULL,
                                        input.map.distance = "cM")

# recommended to save these files for later use
readr::write_rds(my_genotypes, "my_genotypes.rds")

# recommended to save these files for later use
my_parameters <- getIBDparameters(ped.genotypes = my_genotypes,
                                  number.cores = 8)
readr::write_rds(my_parameters, "ibd_parameters.rds")
my_parameters <- readr::read_rds("ibd_parameters.rds")

# recommended to save these files for later use
my_posterior <- getIBDposterior(ped.genotypes = my_genotypes,
                                parameters = my_parameters,
                                error=0.001,
                                number.cores = 8)
readr::write_rds(my_posterior, "ibd_posterior.rds")
my_posterior <- readr::read_rds("ibd_posterior.rds")

# recommended to save these files for later use
my_ibd_1 <- getIBDsegments(ped.genotypes = my_genotypes,
                           parameters = my_parameters,
                           minimum.snps = 1,
                           minimum.length.bp = 1,
                           error = 0.001)
readr::write_rds(my_ibd_1, path="ibd_segments_nsnp_1_len_1.rds")
my_ibd <- readr::read_rds("ibd_segments_nsnp_1_len_1.rds")

# recommended to save these files for later use
my_ibd_20 <- getIBDsegments(ped.genotypes = my_genotypes,
                           parameters = my_parameters,
                           minimum.snps = 20,
                           minimum.length.bp = 20,
                           error = 0.001)

```












## Summary

It's a good idea to round off with a summary of what we did.

PS, note that the name of the tutorial on the PGEforge website can be whatever you want - it does not have to match the name of this tutorial.
