---
title: "Give this whatever name you want"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## The data
For this analysis, we will be using VCF generated from whole genome sequencing data from Vietnam. This data is part of the [pf3k dataset](https://www.malariagen.net/parasite/pf3k) and SNP and Indels. It is also available in [main Data tab](https://mrc-ide.github.io/PGEforge/website_docs/data_description.html) for further details.

Remember to load rehh and dependent R packages before running the analysis:
```{r, warning=FALSE, message=FALSE, error=FALSE}
library(rehh)
library(R.utils)
library(vcfR)
library(data.table)

```


## Estimate EHH and iHS along chromosome 13 

### Convert data from VCF to an object of class haplohh for  chromosome 13

chr13_hh<-data2haplohh(hap_file = "SNP_INDEL_Pf3D7_ALL_v3.combined.filtered.vqslod6.biallelic_snp.Vietnam.vcf.gz", min_perc_geno.mrk = 90, vcf_reader = "data.table", verbose = T, polarize_vcf = F, min_maf=0.005,remove_multiple_markers=T, chr.name= "Pf3D7_13_v3")

### Compute iHH (integrated EHH), iES (integrated EHHS) and inES (integrated normalized EHHS) over chromosome 13

chr13.res <- scan_hh(chr13_hh,discard_integration_at_border = F,threads = 24)

### Compute iHS and save the results for plotting

chr13.ihs<-ihh2ihs(chr13.res, freqbin = 1, verbose = F, standardize = T)
write.table(x=chr13.ihs$ihs, file = "Vitenam_ihs_chr13.csv")

## Summary

It's a good idea to round off with a summary of what we did.

PS, note that the name of the tutorial on the PGEforge website can be whatever you want - it does not have to match the name of this tutorial.
