---
title: "Insert tool title"
output: html_document
author: "Insert name"
date: "Insert todays date"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(kableExtra)
```

<br/>
<br/>

## Summary sheet

```{r, echo=FALSE, warning=FALSE}
# read in summary sheet from file
df_summary <- read.csv("summary_sheet.csv", header = FALSE)

# produce formatted table
df_summary |>
  kbl(format = "html", col.names = NULL, table.attr = "style='width:50%;'", escape = FALSE) |>
  kable_styling(position = "left", bootstrap_options = c("striped", "hover", "condensed"))
```

## Purpose

This tool computes several metrics of positive selection signatures along the genome of an organism. These metrics include the Extended Haplotype Homozygosity (EHH), integrated EHH (IHH), site-specific EHH (EHHS), integrated EHHS (iES), integrated haplotype homozygosity score (iHS), ratio of EHHS (Rsb) and cross-population EHH (XP-EHH). EHH is defined as the probability that two randomly chosen chromosomes, carrying the core allele, are homozygous over a given surrounding chromosomal region. More details on the EHH can be found in the paper published by [Sabeti et al. 2002](https://www.nature.com/articles/nature01140). This istechnically a useful gemomic metrics in malaria, as the nature and length of shared haplotypes and EHH scores flanking an allele, such as a drug resistance mutation, show the ancestral relationship between mutants and how selective sweep survived recombinations over time.

## Existing resources

- This [tutorial](https://cran.r-project.org/web/packages/rehh/vignettes/rehh.html#computing-ihs-rsb-and-xp-ehh) give details on how to run the functions

## Citation


```
 @Article{,
    author = {Mathieu Gautier and Renaud Vitalis},
    title = {rehh: An R package to detect footprints of selection in
      genome-wide SNP data from haplotype structure},
    journal = {Bioinformatics},
    year = {2012},
    volume = {28},
    number = {8},
    pages = {1176-1177},
  }
  @Article{,
    author = {Mathieu Gautier and Alexander Klassmann and Renaud
      Vitalis},
    title = {rehh 2.0: a reimplementation of the R package rehh to
      detect positive selection from haplotype structure},
    journal = {Molecular Ecology Resources},
    year = {2017},
    volume = {17},
    number = {1},
    pages = {78-90},
    doi = {10.1111/1755-0998.12634},
  }

```
